packages/payload/src/auth/strategies/local/incrementLoginAttempts.ts

```diff
@@ -1,59 +1,155 @@
+      },
+    })
+
+    updatedLockUntil = updatedUser.lockUntil
+    updatedLoginAttempts = updatedUser.loginAttempts
   }
 
-  const data: JsonObject = {
-    loginAttempts: Number(doc.loginAttempts) + 1,
+  if (updatedLoginAttempts === null) {
+    throw new Error('Failed to update login attempts or lockUntil for user')
   }
 
-  // Lock the account if at max attempts and not already locked
-  if (typeof doc.loginAttempts === 'number' && doc.loginAttempts + 1 >= maxLoginAttempts) {
-    const lockUntil = new Date(Date.now() + lockTime).toISOString()
-    data.lockUntil = lockUntil
+  // Check updated latest lockUntil and loginAttempts in case there were parallel updates
+  const reachedMaxAttemptsForCurrentUser =
+    typeof updatedLoginAttempts === 'number' && updatedLoginAttempts - 1 >= maxLoginAttempts
+
+  const reachedMaxAttemptsForNextUser =
+    typeof updatedLoginAttempts === 'number' && updatedLoginAttempts >= maxLoginAttempts
+
+  if (reachedMaxAttemptsForCurrentUser) {
+    user.lockUntil = updatedLockUntil
   }
+  user.loginAttempts = updatedLoginAttempts - 1 // -1, as the updated increment is applied for the *next* login attempt, not the current one
+
+  if (
+    reachedMaxAttemptsForNextUser &&
+    (!updatedLockUntil || !isUserLocked(new Date(updatedLockUntil)))
+  ) {
+    // If lockUntil reached max login attempts due to multiple parallel attempts but user was not locked yet,
+    const newLockUntil = new Date(currentTime + lockTime).toISOString()
+
+    await payload.db.updateOne({
+      id: user.id,
+      collection: collection.slug,
+      data: {
+        lockUntil: newLockUntil,
+      },
+      returning: false,
+    })
 
-  await payload.update({
-    id: doc.id,
-    collection: collection.slug,
-    data,
-    depth: 0,
-    req,
-  })
+    if (reachedMaxAttemptsForCurrentUser) {
+      user.lockUntil = newLockUntil
+    }
+
+    if (collection.auth.useSessions) {
+      // Remove all active sessions that have been created in a 20 second window. This protects
+      // against brute force attacks - example: 99 incorrect, 1 correct parallel login attempts.
+      // The correct login attempt will be finished first, as it's faster due to not having to perform
+      // an additional db update here.
+      // However, this request (the incorrect login attempt request) can kill the successful login attempt here.
+
+      // Fetch user sessions separately (do not do this in the updateOne select in order to preserve the returning: true db call optimization)
+      const currentUser = await payload.db.findOne<TypedUser>({
+        collection: collection.slug,
+        select: {
+          sessions: true,
+        },
+        where: {
+          id: {
+            equals: user.id,
+          },
+        },
+      })
+      if (currentUser?.sessions?.length) {
+        // Does not hurt also removing expired sessions
+        currentUser.sessions = removeExpiredSessions(currentUser.sessions).filter((session) => {
```

Comment:
Isn't calling `removeExpiredSessions` redundant here? I think if you're ever in `incrementLoginAttempts` you also would have already called `removeExpiredSessions` inside of `login`. No?